### PR TITLE
Adding optimisation for findtrigs

### PR DESCRIPTION
**Pull request by connor-mcisaac**
While running a search for lensed images we searched all of O2 with a single template using a single detector threshold of SNR=4. This produced roughly 1,000,000 single detector triggers in each detector with the single template.
This means that there were roughly 1,000,000,000,000 coincident triggers in the background to be tested. This obviously takes a long time.

However, when decimation is included we don't need the coincident statistics for most of these as they are removed by decimation. This script can be made more efficient by not calculating a lot of these.

To avoid calculating a lot of these discarded values we can:

1) Find all foreground coincidences and the background for timeslides that will always be kept after decimation.

2) Only keep the foreground triggers and the background that is below the decimation threshold.

3) Take all the single detector triggers in the first detector and calculate the minimum/maximum (depending on the ranking statistic) single detector statistic required in the second detector to pass the decimation threshold.

4) Find the list of all the coincident triggers between the two detectors.

5) Make a list of coincident triggers where the single detector statistic from the second detector passes the maximum/minimum calculated earlier.

6) Only calculate the coincident statistic for the triggers in this list.

7) Keep all background triggers above the decimation threshold.

To extend this to multiple decimation cuts you can loop through steps 3 to 7 starting with the smallest threshold and highest decimation factor.
In step 4 only find coincidences in timeslides that will only keep triggers with coincident statistics above threshold n and below threshold n+1
Then in step 7 only keep triggers that are above threshold n and below threshold n+1

bin/hdfcoinc/pycbc_coinc_findtrigs has been changed to implement this method

bin/hdfcoinc/pycbc_multiifo_coinc_findtrigs has a couple addition changes to deal with 3+ detectors:
After step 2 above find coincidences between all but the pivot ifo with 0 timeslides.
Use the list of coincidences produced by this to find the maximum/minimum statistic needed in the pivot ifo to pass the decimation threshold.

You can then continue in the same way as before but treating these sets of triggers in all the non-pivot detectors like a single trigger.

To calculate the maximum/minimum required single detector statistic methods have been added to pycbc/events/stat.py

coinc_lim_for_thresh
This takes an array of single detector triggers from one detector and a threshold. It then returns the maximum/minimum statistic required in the second detector to pass the threshold.

coinc_multiifo_lim_for_thresh
This takes a list of single detector triggers from multiple ifos (in the same format as coinc_multiifo), a threshold, and the name of the ifo that you want to find the limit for. It then returns the maximum/minimum statistic required in the named detector to pass the threshold.

An attribute "single_increasing" has also been added that is True if a larger single detector statistic produces a larger coincident statistic, and False otherwise. This is convenient for knowing if the limit you get from the previous methods is a maximum or minimum.

So far this has produced the same results as the old scripts when tested, but I need to test for all the different statistics and test for any edge cases that might of been missed.

Also need to be cleaned up/commented properly.

### CODE REVIEW

**Comment by ahnitz** [comment]
@connor-mcisaac The part that seem tricky here is step (3), i.e. deciding which singles can be excluded based on one detector's value. This must be fairly detection statistic specific, though I can imagine that one option is simply to leave this up to the user to decide on valid values for the statistic they've chosen to use. However, even if they've chosen an "unsafe" value, I think if the logic is applied identically to foreground triggers, the results will still be self-consistent; they might just miss some signals. Should this be true, and if so, have you tested it? (i.e. by upping the threshold and seeing if zerolag / background still seem completely consistent with each other).

**Comment by connor-mcisaac** [comment]
@ahnitz Just to make sure this is clear, the thresholds chosen by the user are still in terms of the coincident statistic. (So the choice of decimation factors remains exactly the same as before)

The methods I have added to the ranking statistic classes then convert this into a limit on the single detector statistic.

For example for snr it just returns:
Ifo2lim = Threshold**2 - ifo1stat**2

In the case of things like the phaseTD statistic you can assume that the coincident trigger will have the maximum signal rate possible from the histogram and again, rearrange the equation, get something like:
Ifo2lim = Threshold - ifo1stat - histmax

This is very statistic dependent, which is why I chose to implement it as a method for the ranking statistic classes.

So far the only tests I have done is running some coinc jobs for 2 and 3 detectors using the current code and this PR and making sure the outputs are consistent. I haven't found any cases where they aren't so far.

Sorry if I've misunderstood your question at all.

**Comment by GarethDaviesGW** [comment]
So from what I'm getting - you just assume "best-case scenario" for anything that _isn't_ just a combination of single-detector statistic and use that to reverse-engineer the cutoff sngl_stat?

**Comment by connor-mcisaac** [comment]
Yes exactly

**Comment by ahnitz** [comment]
@connor-mcisaac No, thanks, that all makes sense.

**Comment by GarethDaviesGW** [comment]
Looking through the code it all looks good  (without doing the maths myself to check) - do you have any stats on how much of an improvement in efficiency this is?

**Comment by spxiwh** [comment]
I'll note that while motivated based on our lensing work, this has applications in the standard workflow. In an example test I saw that this makes HV coinc jobs run a factor of 5 faster, while producing the same output (trigger order was a bit different, but was the same content).

**Comment by ahnitz** [comment]
@connor-mcisaac I think the general structure of this change is fine. Did you want us to review this as is in detail now? (wasn't clear if this was a work in progress or not)

**Comment by connor-mcisaac** [comment]
I have a couple of changes to make and then it can be reviewed. I'll hopefully get them done tomorrow then let you know.

**Comment by connor-mcisaac** [comment]
Sorry for the delay, I think this is good to look over now.
@GarethDaviesGW In terms of speed up it only makes a small difference for stats like newsnr (up to a factor of 2 faster)
but for stat like phasetd_exp_fit_stat_sgveto it gets roughly a factor of 5, after that it's limited mostly by reading the data.

File: bin/hdfcoinc/pycbc_coinc_findtrigs [hunk 1]
```diff
@@ -239,94 +252,153 @@ for tnum in template_ids:
     logging.info('Calculating Single Detector Statistic')
     s0g, s1g = rank_method.single(trigs0), rank_method.single(trigs1)
 
+    try:
+        s0gstat = s0g['snglstat'].copy()
```
Comment (GarethDaviesGW) [inline bin/hdfcoinc/pycbc_coinc_findtrigs:256]: When would this work/not work?

File: bin/hdfcoinc/pycbc_coinc_findtrigs [hunk 2]
```diff
@@ -239,94 +252,153 @@ for tnum in template_ids:
...
+        s1lower = {}
+        for kidx in range(1, len(threshes)):
+            # For each trigger in detector 0 get the limit in detector 1 to pass
+            # the current decimation threshold
+            s1lims[kidx] = rank_method.coinc_lim_for_thresh(s0, threshes[kidx])
+            if not rank_method.single_increasing:
+                s1lims[kidx] *= -1.
+            # subtract small amount to account for errors due to rounding
+            s1lims[kidx] -= 1e-6
+            # Get the minimum statistic required for all triggers
```
Comment (GarethDaviesGW) [inline bin/hdfcoinc/pycbc_coinc_findtrigs:296]: maybe clarify to "...for all triggers at this decimation threshold"

File: bin/hdfcoinc/pycbc_coinc_findtrigs [hunk 3]
```diff
@@ -239,94 +252,153 @@ for tnum in template_ids:
...
+            try:
+                s1stat = s1['snglstat'].copy()
+            except IndexError:
+                s1stat = s1.copy()
+
+            if not rank_method.single_increasing:
+                s1stat *= -1.
+
+            tidx = len(threshes)
+            for i in range(1, len(threshes)):
```
Comment (GarethDaviesGW) [inline bin/hdfcoinc/pycbc_coinc_findtrigs:345]: a comment for explanation of what is happening in this loop would be good - it took me a few minutes to work out what was happening here

File: bin/hdfcoinc/pycbc_multiifo_coinc_findtrigs [hunk 4]
```diff
@@ -182,96 +193,206 @@ for tnum in template_ids:
...
         continue
 
+    if type(rank_method.single_dtype) == list:
+        entries = [e[0] for e in rank_method.single_dtype]
+        if 'snr' in entries:
+            min_snr = min([numpy.min(sds_full[i]['snr']) for i in trigs.ifos])
+        else:
+            min_snr = None
+        if 'sigmasq'in entries:
+            max_sigmasq = max([numpy.max(sds_full[i]['sigmasq']) for i in trigs.ifos])
```
Comment (GarethDaviesGW) [inline bin/hdfcoinc/pycbc_multiifo_coinc_findtrigs:203]: This is slightly wrong in terms of actually getting the maximum possible network SNR (which is what this is used for later); you should be doing min() of the maxes in the list

File: bin/hdfcoinc/pycbc_coinc_findtrigs (hunk 1 above)
Comment (connor-mcisaac) [reply bin/hdfcoinc/pycbc_coinc_findtrigs:256]: This wouldn't work if you're using newsnr, as in that case s0g is just a singe array.

Comment (GarethDaviesGW) [reply bin/hdfcoinc/pycbc_coinc_findtrigs:256]: Okay, would you be able to add a comment to say that "s0g is sometimes an array of arrays, sometimes just a single array"

Comment (connor-mcisaac) [reply bin/hdfcoinc/pycbc_coinc_findtrigs:257]: A comment has been added to explain this

File: bin/hdfcoinc/pycbc_coinc_findtrigs (hunk 3 above)
Comment (connor-mcisaac) [reply bin/hdfcoinc/pycbc_coinc_findtrigs:350]: Does this comment read okay?

File: bin/hdfcoinc/pycbc_coinc_findtrigs (hunk 2 above)
Comment (connor-mcisaac) [reply bin/hdfcoinc/pycbc_coinc_findtrigs]: This comment has been updated

File: bin/hdfcoinc/pycbc_multiifo_coinc_findtrigs (hunk 4 above)
Comment (connor-mcisaac) [reply bin/hdfcoinc/pycbc_multiifo_coinc_findtrigs]: This has been fixed now

File: bin/hdfcoinc/pycbc_coinc_findtrigs (hunk 3 above)
Comment (GarethDaviesGW) [reply bin/hdfcoinc/pycbc_coinc_findtrigs:350]: looks good to me
